### PR TITLE
fix: execution fails when istio CRD is not installed in the cluster

### LINKED ISSUE #138: Execution fails when CRDs are not installed in the cluster
**What happened**:
When running the tool without `--providers` flag, it defaults to all providers. Then, when istio provider attempts to read its CRs, it fails the execution

**What you expected to happen**:
If I am not specifying istio as the desired provider, execution should work regardless of whether I have the CRDs installed or not

**How to reproduce it (as minimally and precisely as possible)**:
Run `ingress2gateway print` in a cluster without istio CRDs intalled.

Here is the error
```
./ingress2gateway print
Error: failed to read istio resources from the cluster: failed to read resources from cluster: failed to read gateways: failed to list istio gateways: failed to get API group resources: unable to retrieve the complete list of server APIs: networking.istio.io/v1beta1: the server could not find the requested resource
```

Two suggestion in mind for how to address this:
1) Enforce to specify `--providers`, and not default to all providers
2) Fail soft with a warning if the provider specified is not the provider whose CRDs are not installed

/assign @dpasiukevich 
/cc @levikobi 
/cc @mlavacca

### PR DESCRIPTION
**Pull request by YTGhost**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
1. Enforce to specify --providers, and not default to all providers
2. Fail soft with a warning if the provider specified is not the provider whose CRDs are not installed

**Which issue(s) this PR fixes**:

Fixes #138 

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by YTGhost** [comment]
@dpasiukevich @LiorLieberman PTAL

File: cmd/print.go [hunk 1]
```diff
@@ -65,9 +65,12 @@ type PrintRunner struct {
 // converted Gateways and HTTP Routes. The steps include reading from the source,
 // construct ingresses, convert them, then print them out.
 func (pr *PrintRunner) PrintGatewaysAndHTTPRoutes(cmd *cobra.Command, _ []string) error {
+	if pr.providers == nil || len(pr.providers) == 0 {
+		return fmt.Errorf("you must specify at least one provider with --providers")
+	}
```
Comment (LiorLieberman) [inline cmd/print.go:70]: consider using [MarkFlagRequired](https://pkg.go.dev/github.com/spf13/cobra#MarkFlagRequired) instead

Comment (YTGhost) [reply cmd/print.go:70]: That's a good idea, thanks!

**Review by LiorLieberman — COMMENTED** [review]
Thanks, left some comments. 
A general comment, you made two proposed solutions to the same problem. if we change the code to fail soft we might not need to enforce specifying providers.

File: pkg/i2gw/providers/istio/resource_reader.go [hunk 2]
```diff
@@ -44,6 +45,12 @@ func newResourceReader(conf *i2gw.ProviderConf) reader {
 func (r *reader) readResourcesFromCluster(ctx context.Context) (*storage, error) {
 	res := newResourcesStorage()
 
+	crd := &apiextensionsv1.CustomResourceDefinition{}
+	if err := r.conf.Client.Get(ctx, types.NamespacedName{Name: CRDName}, crd); err != nil {
+		log.Printf("CRD '%s' is not installed in the cluster, will not continue reading resources", CRDName)
+		return res, nil
+	}
+
```
Comment (LiorLieberman) [inline pkg/i2gw/providers/istio/resource_reader.go:53]: Sorry this was not added to the previous review. 
The intention was to change the call to read ANY CRD. In this case you only cover Istio Gateway but it is also relevant for VirtualServices and any other provider's CRD that will be added in the future.

So either you change istio calls to "fail soft" or you create a general helper that doing it

**Review by mlavacca — COMMENTED** [review]
I lean toward not requiring any provider when calling the CLI, instead, I propose the fail soft as @LiorLieberman proposed. I think the soft fail logic should be introduced one level abve. If a provider fails to read resources, we might want to just log the error and continue with other providers. We should implement this outside the istio provider as we might have the same issue with other Providers (Kong just implemented `TCPIngress`es conversion, and I think we have the same issue there). 

At the moment, logging messages is problematic because of https://github.com/kubernetes-sigs/ingress2gateway/issues/145, but that's a separate issue.

**Review by levikobi — COMMENTED** [review]
+1 to what @mlavacca and @LiorLieberman said. I also would not want to see an enforcement of specifying the `providers` flags, as this would be one more friction point that users have to figure out.

Alongside Kong, Apisix also faces this issue now. We need to solve this problem for all the providers (one level above, again, as @mlavacca said).

I think a reasonable solution is to have each provider check if his CRDs are installed; if they're not, return a typed error that specifies that. This error will result in a "fail soft" behavior one level above.

At the end of the run, we can specify which providers participated in the translation instead of describing which didn't. I think that's a better choice since a naive user who has some manifests but doesn't specify providers shouldn't receive multiple error messages about missing CRDs from different providers (which he might not even be familiar with).

**Comment by LiorLieberman** [comment]
@YTGhost This is a very important fix to add. 
Do you still have the bandwidth to work on this?

**Comment by levikobi** [comment]
> This is a very important fix to add. 

+1

**Comment by YTGhost** [comment]
> @YTGhost This is a very important fix to add. Do you still have the bandwidth to work on this?

@LiorLieberman Sorry for the delay, Let me fix it tonight.

And let me confirm what we need to do now: 
- [ ] Not enforcing the specification of `--providers`
- [ ] Implement a general helper that checks whether the list of CRDs passed to it are installed in the cluster. If not installed, it will return a typed error. For this typed error, we will use a soft fail logic to handle it at a higher level.
- [ ] Introduce this helper to check each current provider.

**Comment by mlavacca** [comment]
> > @YTGhost This is a very important fix to add. Do you still have the bandwidth to work on this?
> 
> @LiorLieberman Sorry for the delay, I will fix it as soon as possible.
> 
> And let me confirm what we need to do now:
> 
> * [ ]  Not enforcing the specification of `--providers`

👍 

> * [ ]  Implement a general helper that checks whether the list of CRDs passed to it are installed in the cluster. If not installed, it will return a typed error. For this typed error, we will use a soft fail logic to handle it at a higher level.

I don't think we need a helper to check the CRDs are installed in the cluster. This is the flow I have in mind:
-  The providers' `readResourcesFromCluster` functions returns a typed error when the CRD is not installed in the cluster (instead of a general error)
- the [caller](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/ingress2gateway.go#L93-L100) of `readResourcesFromCluster` checks the error type. If it is related to missing CRDs, we add the provider to the list of the ones who did not provide translation, as @levikobi suggested.

**Comment by LiorLieberman** [comment]
@YTGhost can we help to speed this up?
It is also ok if you have had limited bandwidth recently, just that this feature is really important so if you did, someone else would implement this.

**Comment by YTGhost** [comment]
> @YTGhost can we help to speed this up? It is also ok if you have had limited bandwidth recently, just that this feature is really important so if you did, someone else would implement this.

Yes, sorry, I've been busy with some school matters recently, and don't have enough bandwidth. It would be great if someone could help implement it. Apologies once again.

**Comment by mlavacca** [comment]
closing as superseded by #153

**Pull request closed without merging**